PTP: ENET NXP K64 timestamping also PTP 1588  messages.

theadib opened the pull request:
add time-stamping not only to gPTP VLAN
make sure second field is filled

theadib committed:
drivers: ethernet: nxp: timestamping to all potential packets.

add timestamping on Tx to packets marked for timestamping
add timestamping on Rx to all packets for later use
fix race condidition on adding timestamp when sending delay_req

Signed-off-by: Adib Taraben <theadib@gmail.com>

theadib commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -396,7 +413,8 @@ static int eth_nxp_enet_rx(const struct device *dev)
 	/* Timestamp the packet using PTP clock */
-	if (eth_get_ptp_data(get_iface(data), pkt)) {
+	if (eth_get_ptp_data(get_iface(data), pkt) || 1) {
+		/* should timestamp at least all PTP VLAN (gPTP) and PTP packet (1588) packets */
```
in eth_get_ptp_data() there is this evaluation:
```
	if (net_eth_is_vlan_enabled(eth_ctx, iface)) {
		pkt_is_ptp = ntohs(hdr_vlan->type) == NET_ETH_PTYPE_PTP;
	} else {
		pkt_is_ptp = ntohs(NET_ETH_HDR(pkt)->type) == NET_ETH_PTYPE_PTP;
	}
```	
hence the reference to VLAN. 
The packet type evaluated is no standard ETH IP frame. But some PTP for gPTP and Ethernet AVS related.

jukkar replied:
The commit def4a7ec1299c5b1e844858aa865555bf33f905e removed VLAN support from gPTP. It seems that this driver has still some checks for it, those could be removed in a separate commit.

yangbolu1991 replied:
In my opinion, timestamping is not absolute ptp specific.

Ideally, like linux, timestamping should be controlled via application(socket/ioctl).
SO_TIMESTAMPING flags
https://github.com/torvalds/linux/blob/master/include/uapi/linux/net_tstamp.h#L27
hwtstamp_tx_types
https://github.com/torvalds/linux/blob/master/include/uapi/linux/net_tstamp.h#L112
hwtstamp_rx_filters
https://github.com/torvalds/linux/blob/master/include/uapi/linux/net_tstamp.h#L150

But in zephyr, currently timestamping used by gptp and ptp. ptp is using SO_TIMESTAMPING flags while gptp doesn't.
So, for TX, the changes look ok for both ptp/gptp support.
```
eth_get_ptp_data(net_pkt_iface(pkt), pkt) || net_pkt_is_tx_timestamping(pkt)
```

for RX, may be as below?

```
eth_get_ptp_data(net_pkt_iface(pkt), pkt) || net_pkt_is_rx_timestamping(pkt)
```

Actually, for RX, I would prefer directly to give all packets timestamps, if there are not any hardware filter on mac.
@jukkar  What do you think about this?
Thanks.

yangbolu1991 replied:
Can we just remove the if condition?
Timestamping all RX packet is the hardware capability, application could decide to use or not.

theadib commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -164,11 +164,25 @@ static inline void ts_register_tx_event(const struct device *dev,
 	struct net_pkt *pkt = frameinfo->context;
```
done

hakehuang commented:
this fixes works on my frdm_k64f board with below:

```

uart:~$ net gptp
Port Interface          Role
 1   0x20000a68 [1]     SLAVE

SiteSyncSync state machine variables:
        Current state                  : RECEIVING_SYNC
        A PortSyncSync struct is ready : no
ClockSlaveSync state machine variables:
        Current state                  : SEND_SYNC_IND
        A PortSyncSync struct is ready : no
        The local clock has expired    : no
PortRoleSelection state machine variables:
        Current state                  : ROLE_SELECTION
ClockMasterSyncReceive state machine variables:
        Current state                  : SOURCE_TIME
        A ClockSourceTime              : yes
        The local clock has expired    : no
uart:~$ net gptp 1
Port id    : 1 (SLAVE)
Interface  : 0x20000a68 [1]
Clock id   : 02:04:9f:ff:fe:37:a9:34
Version    : 2
AS capable : yes

Configuration:
Time synchronization and Best Master Selection enabled        : yes
The port is measuring the path delay                          : yes
One way propagation time on the link attached to this port    : 1760 ns
Propagation time threshold for the link attached to this port : 1000000 ns
Estimate of the ratio of the frequency with the peer          : 0
Asymmetry on the link relative to the grand master time base  : 0
Maximum interval between sync messages                        : 375000000
Maximum number of Path Delay Requests without a response      : 3
Current Sync sequence id for this port                        : 8954
Current Path Delay Request sequence id for this port          : 2350
Current Announce sequence id for this port                    : 41134
Current Signaling sequence id for this port                   : 60906
Whether neighborRateRatio needs to be computed for this port  : yes
Whether neighborPropDelay needs to be computed for this port  : yes
Initial Announce Interval as a Logarithm to base 2            : 0
Current Announce Interval as a Logarithm to base 2            : 0
Initial Sync Interval as a Logarithm to base 2                : -4
Current Sync Interval as a Logarithm to base 2                : -4
Initial Path Delay Request Interval as a Logarithm to base 2  : 0
Current Path Delay Request Interval as a Logarithm to base 2  : 0
Time without receiving announce messages before running BMCA  : 3000 ms (3)
Time without receiving sync messages before running BMCA      : 0 ms (3)
Sync event transmission interval for the port                 : 62 ms
Path Delay Request transmission interval for the port         : 1000 ms
BMCA default priority1                                        : 248
BMCA default priority2                                        : 248

Runtime status:
Current global port state                                : SLAVE
Path Delay Request state machine variables:
        Current state                                    : WAIT_ITV_TIMER
        Initial Path Delay Response Peer Timestamp       : 301245205200
        Initial Path Delay Response Ingress Timestamp    : 301245206275
        Path Delay Response messages received            : 1
        Path Delay Follow Up messages received           : 1
        Number of lost Path Delay Responses              : 0
        Timer expired send a new Path Delay Request      : 0
        NeighborRateRatio has been computed successfully : 1
        Path Delay has already been computed after init  : 0
        Count consecutive reqs with multiple responses   : 0
Path Delay Response state machine variables:
        Current state                                    : INITIAL_WAIT_REQ
SyncReceive state machine variables:
        Current state                                    : WAIT_SYNC
        A Sync Message has been received                 : no
        A Follow Up Message has been received            : no
        A Follow Up Message timeout                      : no
        Time at which a Sync Message without Follow Up
                                     will be discarded   : 0
SyncSend state machine variables:
        Current state                                    : SEND_SYNC
        A MDSyncSend structure has been received         : no
        The timestamp for the sync msg has been received : no
PortSyncSyncReceive state machine variables:
        Current state                                    : RECEIVED_SYNC
        Grand Master / Local Clock frequency ratio       : 1.000000
        A MDSyncReceive struct is ready to be processed  : no
        Expiry of SyncReceiptTimeoutTimer                : no
PortSyncSyncSend state machine variables:
        Current state                                    : SYNC_RECEIPT_TIMEOUT
        Follow Up Correction Field of last recv PSS      : 0
        Upstream Tx Time of the last recv PortSyncSync   : 839502794640
        Rate Ratio of the last received PortSyncSync     : 1.000000
        GM Freq Change of the last received PortSyncSync : 0.000000
        GM Time Base Indicator of last recv PortSyncSync : 0
        Received Port Number of last recv PortSyncSync   : 0
        PortSyncSync structure is ready to be processed  : yes
        Flag when the half_sync_itv_timer has expired    : yes
        Has half_sync_itv_timer expired twice            : no
        Has syncReceiptTimeoutTime expired               : yes
PortAnnounceReceive state machine variables:
        Current state                                    : RECEIVE
        An announce message is ready to be processed     : no
PortAnnounceInformation state machine variables:
        Current state                                    : CURRENT
        Expired announce information                     : no
PortAnnounceTransmit state machine variables:
        Current state                                    : POST_IDLE
        Trigger announce information                     : no

Statistics:
Sync messages received                 : 1760
Follow Up messages received            : 1974
Path Delay Request messages received   : 220
Path Delay Response messages received  : 220
Path Delay messages threshold exceeded : 0
Path Delay Follow Up messages received : 220
Announce messages received             : 219
ptp messages discarded                 : 0
Sync reception timeout                 : 1
Announce reception timeout             : 0
Path Delay Requests without a response : 832
Sync messages sent                     : 10
Follow Up messages sent                : 10
Path Delay Request messages sent       : 1052
Path Delay Response messages sent      : 220
Path Delay Response FUP messages sent  : 220
Announce messages sent                 : 1

```

hakehuang commented:
just to note, this as before still seeing, after overnight testing

```
Announce messages received             : 360
ptp messages discarded                 : 0
Sync reception timeout                 : 3
Announce reception timeout             : 1
Path Delay Requests without a response : 845
```

theadib commented:
Hello @yangbolu1991 , I saw your commits on the same file, could you doublecheck this PR does not break your setup?

The PR got stale and meanwhile had merge conflicts. I rebased to main.
please @jukkar  @decsny re-evaluate this PR for merging.

Many thanks in advance, Adib

yangbolu1991 commented:
BTW, I send some fix and improvement on ptp, if you are interested.
https://github.com/zephyrproject-rtos/zephyr/pull/89517

Thanks.

yangbolu1991 commented:
@theadib @jukkar 
Another key fix for ptp.
https://github.com/zephyrproject-rtos/zephyr/pull/89658

Thanks.

theadib commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -163,11 +163,25 @@ static inline void ts_register_tx_event(const struct device *dev,
+
 			pkt->timestamp.nanosecond = frameinfo->timeStamp.nanosecond;
-			pkt->timestamp.second = frameinfo->timeStamp.second;
+			pkt->timestamp.second = ptp_time.second;
```
Hi @yangbolu1991  your change **drivers: ptp_clock_nxp_enet: avoid configuring IRQ handlers again** 
**creates a regression on the K64 platform.** :-(
There is No PTP interrupt enabled at all and so the Seconds field is not updated on the overflow interrupt.
the K64 platform has no hardware PTP seconds counter.

my hack is to place a enable_ISR rigth after start_timer
```
diff --git a/drivers/ptp_clock/ptp_clock_nxp_enet.c b/drivers/ptp_clock/ptp_clock_nxp_enet.c
index 5165873fca9..45acb214d2f 100644
--- a/drivers/ptp_clock/ptp_clock_nxp_enet.c
+++ b/drivers/ptp_clock/ptp_clock_nxp_enet.c
@@ -180,6 +180,8 @@ void nxp_enet_ptp_clock_callback(const struct device *dev,
                ENET_Ptp1588SetChannelMode(data->base, kENET_PtpTimerChannel3,
                                kENET_PtpChannelPulseHighonCompare, true);
                ENET_Ptp1588StartTimer(data->base, ptp_config.ptp1588ClockSrc_Hz);
+               ENET_EnableInterrupts(data->base, ((uint32_t)kENET_TsTimerInterrupt | (uint32_t)kENET_TsAvailInterrupt));
+
        }
 }
```

yangbolu1991 replied:
I will suggest to add below.
Because kENET_TsAvailInterrupt event was actually never used. The TX timestamp was got from TX event handler.

```
ENET_EnableInterrupts(data->base, ((uint32_t)kENET_TsTimerInterrupt);
```

I guess you had verified it on K64? If so, I think I found a another problem.
Currently kENET_TsTimerInterrupt handler ENET_TimeStampIRQHandler is via ptp_clock_nxp_enet_isr (tstimer interrupt source). While actually this should be via eth_nxp_enet_isr (enet interrupt source).
The ENET_TimeStampIRQHandler should be moved.

Previously all worked fine on RT1060 was because tstimer interrupt source was triggered frequently by time comparing event.

So, I'd like to know if you had verified on K64.
If so, the interrupt source of kENET_TsTimerInterrupt is different between RT1060 and K64.

Thanks.

theadib replied:
Hi @yangbolu1991 yes my platform is the FRDM_K64F using MK64FN1MOVLL12.
I test on that board. I do not have RT1060 or others.

Please let me know how you adapt to that regression. 
IMHO a double initialization is acceptable. :-) If it works.
regards, Adib

yangbolu1991 replied:
@theadib I created a PR.
https://github.com/zephyrproject-rtos/zephyr/pull/90117

Could you help to try if that works for K64? I guess no need to fill the seconds with it.
Thanks.

theadib commented:
Hi @yangbolu1991 I rebased this PR to what is current main.

The current main does not update the PTP timer,
When receiving PTP protocol there is a big jump in PTP time.

I face in my branch a BUS FAULT now :-( the problem is in the tx_tstamp thread. 
Do you have any idea where that comes from? Do you face similar issue?

```
[00:00:09.635,000] <dbg> ptp_port: ptp_port_timer_event_gen: Port 1 RX Sync timeout
[00:00:09.635,000] <dbg> ptp_port: port_state_update: Port 1 changed state from TIME RECEIVER to TIME TRANSMITTER
[00:00:10.635,000] <dbg> ptp_port: ptp_port_timer_event_gen: Port 1 TX Sync timeout
[00:00:10.635,000] <dbg> ptp_port: port_sync_msg_transmit: Port 1 sends Sync message
[00:00:10.636,000] <dbg> ptp_port: port_sync_timestamp_cb: Port 1 sends Follow_Up message
[00:00:11.635,000] <dbg> ptp_port: ptp_port_timer_event_gen: Port 1 TimeTransmitter Announce timeout
[00:00:11.635,000] <err> os: ***** BUS FAULT *****
[00:00:11.635,000] <err> os:   Imprecise data bus error
[00:00:11.635,000] <err> os: r0/a1:  0x20003f58  r1/a2:  0x20000aa0  r2/a3:  0x00000000
[00:00:11.635,000] <err> os: r3/a4:  0x00000000 r12/ip:  0x20000aa0 r14/lr:  0x0002f55f
[00:00:11.635,000] <err> os:  xpsr:  0x61000000
[00:00:11.635,000] <err> os: Faulting instruction address (r15/pc): 0x0001ede6
[00:00:11.635,000] <err> os: >>> ZEPHYR FATAL ERROR 26: Unknown error on CPU 0
[00:00:11.635,000] <err> os: Current thread: 0x20003f58 (tx_tstamp)
[00:00:11.760,000] <err> os: Halting system
```
Thanks, Adib

theadib commented:
hello @yangbolu1991 @decsny @jukkar 
I aligned the PR to the current main and all @yangbolu1991 patches.
Now it works again with my FRDM_K64F
if you can have a look please
Thanks, Adib.

yangbolu1991 commented on drivers/ethernet/eth_nxp_enet.c:
```diff
@@ -163,12 +163,15 @@ static inline void ts_register_tx_event(const struct device *dev,
-		if (eth_get_ptp_data(net_pkt_iface(pkt), pkt) && frameinfo->isTsAvail) {
+		if ((eth_get_ptp_data(net_pkt_iface(pkt), pkt) ||
+		     net_pkt_is_tx_timestamping(pkt)) &&
+		    frameinfo->isTsAvail) {
```
How about keep only `if (frameinfo->isTsAvail)` ? Because during TX, it had already do the same filtering.

theadib commented:
thanks @yangbolu1991 
I updated the commit message 
I had to update also PTP code - the timestamp could not be added because the packet was not in the list when sending and the after send callback happend before adding the packet to the list 
Potentially my code runs just a bit too slow ;-)

I removed the if(|| 1) query and adding this code unconditionally.

 Regarding other if() relaxations I am a bit scared to break something because I do not run the other tests.
 I would just replace `eth_get_ptp_data()` by  `net_pkt_is_tx_timestamping()` but too scared to break someone else code

maass-hamburg commented:
changed  Assignees as this is mainly NXP stuff
